File: c/1-blinky/main.c
#include <bits/types/siginfo_t.h>
#include <errno.h>
#include <gpiod.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/resource.h>
#include <unistd.h>

#include "memory.h"
#include "perf.h"

const char CONSUMER[] = "Consumer";
const char CHIPNAME[] = "gpiochip0";
const int32_t LINE_NUMBER = 13;

const int64_t BLINK_FREQUENCY = 10;
const int64_t UPDATE_FREQUENCY = BLINK_FREQUENCY * 2;
const int64_t SLEEP_DURATION_US = 1000000 / UPDATE_FREQUENCY;

bool do_continue = true;

void interrupt_handler(int) {
  printf("\nGracefully stopping\n");
  do_continue = false;
}
const struct sigaction interrupt_sigaction = {
    .sa_handler = &interrupt_handler,
};

int main(int, char **) {
  int res;

  printf("Controlling an LED from C\n");

  res = sigaction(SIGINT, &interrupt_sigaction, NULL);
  if (res != 0) {
    fprintf(stderr, "sigaction fail (%d): %s\n", res, strerror(errno));
    return EXIT_FAILURE;
  }

  struct gpiod_chip *chip = gpiod_chip_open_by_name(CHIPNAME);
  if (!chip) {
    fprintf(stderr, "gpiod_chip_open_by_name fail: %s\n", strerror(errno));
    return EXIT_FAILURE;
  }

  struct gpiod_line *line = gpiod_chip_get_line(chip, LINE_NUMBER);
  if (!line) {
    fprintf(stderr, "gpiod_chip_get_line fail: %s\n", strerror(errno));
    gpiod_chip_close(chip);
    return EXIT_FAILURE;
  }

  res = gpiod_line_request_output(line, CONSUMER, 0);
  if (res != 0) {
    fprintf(
        stderr, "gpiod_line_request_output fail (%d): %s\n", res,
        strerror(errno)
    );
    gpiod_line_release(line);
    gpiod_chip_close(chip);
    return EXIT_FAILURE;
  }

  bool is_on = false;

  perf_counter_t *perf;
  res = perf_counter_init(&perf, "MAIN", UPDATE_FREQUENCY * 2);
  if (res != 0) {
    fprintf(stderr, "perf_counter_init fail (%d): %s\n", res, strerror(errno));
    return EXIT_FAILURE;
  }

  while (do_continue) {
    for (size_t i = 0; i < UPDATE_FREQUENCY; ++i) {
      usleep(SLEEP_DURATION_US);

      const perf_mark_t start = perf_mark();

#ifdef DEBUG
      printf("Turning the LED %s", led_state ? "ON" : "OFF");
#endif

      res = gpiod_line_set_value(line, is_on);
      if (res != 0) {
        fprintf(
            stderr, "gpiod_line_set_value fail (%d): %s\n", res, strerror(errno)
        );
        continue;
      }

      is_on = !is_on;

      perf_counter_add_sample(perf, start);
    }

    memory_report();
    perf_counter_report(perf);
    perf_counter_reset(perf);
  }

  perf_counter_deinit(perf);
  res = gpiod_line_set_value(line, 0);
  if (res != 0) {
    fprintf(
        stderr, "gpiod_line_set_value fail (%d): %s\n", res, strerror(errno)
    );
  }
  gpiod_line_release(line);
  gpiod_chip_close(chip);

  return EXIT_SUCCESS;
}

File: zig/1-blinky/main.zig
const std = @import("std");
const config = @import("config");
const gpio = @import("gpio");

const c = @import("c.zig");
const memory = @import("memory.zig");
const perf = @import("perf.zig");

const line_number = 13;

const blink_frequency = 10;
const update_frequency = 2 * blink_frequency;
const sleep_time_ns = std.time.ns_per_s / update_frequency;

var do_continue = true;
pub fn interrupt_handler(_: c_int) callconv(.C) void {
    std.debug.print("\nGracefully stopping\n", .{});
    do_continue = false;
}

pub fn main() !void {
    std.log.info("Controlling an LED from Zig\n", .{});

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    var counting = memory.CountingAllocator.init(gpa.allocator());
    const allocator = counting.allocator();

    const signal = @intFromPtr(c.signal(c.SIGINT, &interrupt_handler));
    if (signal < 0)
        return std.posix.unexpectedErrno(std.posix.errno(signal));

    var chip = try gpio.getChip("/dev/gpiochip0");
    defer chip.close();

    var line = try chip.requestLine(line_number, .{ .output = true });
    defer line.close();

    var perf_main = try perf.Counter.init(allocator, "MAIN", update_frequency * 2);
    defer perf_main.deinit();

    var report_number: u64 = 0;

    var is_on = false;
    while (do_continue) {
        for (0..update_frequency) |_| {
            std.time.sleep(sleep_time_ns);

            const start = perf.Marker.now();

            std.log.debug("Turning the LED {s}", .{if (is_on) "ON" else "OFF"});
            line.setValue(is_on) catch |err| std.log.err("Line.setValue fail: {}", .{err});
            is_on = !is_on;

            perf_main.add_sample(start);
        }

        std.log.info("# REPORT {}", .{report_number});
        memory.report();
        perf_main.report();
        perf_main.reset();
        report_number += 1;
    }
}

pub const std_options: std.Options = .{
    .log_level = std.enums.nameCast(std.log.Level, config.log_level),
};
